autode/opt/coordinates/primitives.py

```diff
@@ -0,0 +1,222 @@
+import numpy as np
+from abc import ABC, abstractmethod
+from typing import Tuple
+
+
+class Primitive(ABC):
+    """Primitive internal coordinate"""
+
+    @abstractmethod
+    def __call__(self,
+                 x: 'autode.opt.coordinates.CartesianCoordinates'
+                 ) -> float:
+        """Return the value of this PIC given a set of cartesian coordinates"""
+
+    @abstractmethod
+    def derivative(self,
+                   i:         int,
+                   component: str,
+                   x:         'autode.opt.coordinates.CartesianCoordinates'
+                   ) -> float:
+        r"""
+        Calculate the derivative with respect to a cartesian coordinate
+
+        .. math::
+
+            \frac{dq}
+                  {d\boldsymbol{X}_{i, k}} {\Bigg\rvert}_{X=X0}
+
+        where :math:`q` is the primitive coordinate and :math:`\boldsymbol{X}`
+        are the cartesian coordinates.
+
+        -----------------------------------------------------------------------
+        Arguments:
+            i: Cartesian index to take the derivative with respect to;
+                0-N for N atoms
+
+            component: Cartesian component (x, y, z) to take the derivative
+                       with respect to. {'x', 'y', 'z'}
+
+            x: Cartesian coordinates
+
+        Returns:
+            (float): Derivative
+        """
+
+    @staticmethod
+    def _cart_component_to_idx(component: str) -> int:
```

Comment:
Could you replace this function with a dictionary?

Reply:
This was a dictionary before(!) then changed it to this. I thought this was more readable than

```python
def _cart_component_to_idx(component: str) -> int:
    return {'x': 0, 'y': 1, 'z': 2}[component]
```

but maybe not?!

Reply:
In my mind you'd get rid of the function call somehow, _i.e._ the call to `_cart_component_to_idx` would be replace by looking up a dictionary. By maybe it doesn't fit you current design well? Also an `enum` (https://docs.python.org/3/library/enum.html) may be even better than a dict. Not a major issue though, I wouldn't spend to much time on this

Reply:
💯 agree. [IntEnum](https://docs.python.org/3/library/enum.html#intenum) seems perfect, and it'll make @josephsilcock happy!